pvlib/iotools/pvgis.py

```diff
@@ -23,6 +23,348 @@
+    elif start is not None:
+        params['startyear'] = start if isinstance(start, int) else start.year
+    if end is not None:
+        params['endyear'] = end if isinstance(end, int) else end.year
+    if peakpower is not None:
+        params['peakpower'] = peakpower
+
+    # The url endpoint for hourly radiation is 'seriescalc'
+    res = requests.get(url + 'seriescalc', params=params, timeout=timeout)
+    # PVGIS returns really well formatted error messages in JSON for HTTP/1.1
+    # 400 BAD REQUEST so try to return that if possible, otherwise raise the
+    # HTTP/1.1 error caught by requests
+    if not res.ok:
+        try:
+            err_msg = res.json()
+        except Exception:
+            res.raise_for_status()
+        else:
+            raise requests.HTTPError(err_msg['message'])
+
+    # initialize data to None in case API fails to respond to bad outputformat
+    data = None, None, None
+    if outputformat == 'json':
+        src = res.json()
+        return _parse_pvgis_hourly_json(src, map_variables=map_variables)
+    elif outputformat == 'csv':
+        with io.StringIO(res.content.decode('utf-8')) as src:
+            return _parse_pvgis_hourly_csv(src, map_variables=map_variables)
+    else:
+        # this line is never reached because if outputformat is not valid then
+        # the response is HTTP/1.1 400 BAD REQUEST which is handled earlier
+        pass
+    return data
+
+
+def _parse_pvgis_hourly_json(src, map_variables):
+    inputs = src['inputs']
+    metadata = src['meta']
+    data = pd.DataFrame(src['outputs']['hourly'])
+    data.index = pd.to_datetime(data['time'], format='%Y%m%d:%H%M', utc=True)
+    data = data.drop('time', axis=1)
+    data = data.astype(dtype={'Int': 'int'})  # The 'Int' column to be integer
+    if map_variables:
+        data.rename(columns=PVGIS_VARIABLE_MAP, inplace=True)
+    return data, inputs, metadata
+
+
+def _parse_pvgis_hourly_csv(src, map_variables):
+    # The first 4 rows are latitude, longitude, elevation, radiation database
+    inputs = {}
+    # 'Latitude (decimal degrees): 45.000\r\n'
+    inputs['latitude'] = float(src.readline().split(':')[1])
+    # 'Longitude (decimal degrees): 8.000\r\n'
+    inputs['longitude'] = float(src.readline().split(':')[1])
+    # Elevation (m): 1389.0\r\n
+    inputs['elevation'] = float(src.readline().split(':')[1])
+    # 'Radiation database: \tPVGIS-SARAH\r\n'
+    inputs['radiation_database'] = str(src.readline().split(':')[1]
+                                       .replace('\t', '').replace('\n', ''))
+    # Parse through the remaining metadata section (the number of lines for
+    # this section depends on the requested parameters)
+    while True:
+        line = src.readline()
+        if line.startswith('time,'):  # The data header starts with 'time,'
+            # The last line of the metadata section contains the column names
+            names = line.strip().split(',')
+            break
```

Comment:
How do we feel about the possibility of an infinite loop if `time,` isn't in the response? Could happen if people are using local files that have been edited or if the API changes. Seems that we should at least break if `line` is empty (end of file). Guessing that'll cause an exception somewhere down the line but could add one here (I don't think it's all that important).

Reply:
I think this is a good idea. Wouldn't it be better to raise an error instead of breaking though (as it would mean that no data section was detected)?

Reply:
@wholmgren How do you feel about raising an IOError?

        elif line == '':  # If end of file is reached
            raise IOError('No data section was detected. File has probably '
                          'been modified since being downloaded from PVGIS')

Reply:
I'm all for raising an exception. I'm not sure about `IOError`. It was [merged into `OSError`](https://docs.python.org/3/library/exceptions.html#OSError) and I think there's discussion of deprecating it in python 4+. I suggest `ValueError` since it relates to a problem with the values within the file/buffer.